### PR TITLE
Issue#87 New Worksheet Button

### PR DESCRIPTION
**Pull request by ChengshuLi**
Mostly done.

To-do:
After I check in chat box PR, I will be able to grab user info and give a correct default worksheet name, such as 'pliang-sampleworksheet' rather than 'user-sampleworksheet'. 

@percyliang

### CODE REVIEW

File: codalab/apps/web/static/js/worksheet/new_worksheet.jsx [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    };
+  },
+
+  componentDidMount: function() {
+  },
+
+  toggleNewWorksheet: function() {
+    if (this.state.showNewWorksheet) {
+      $('#new-worksheet').css('display', 'none');
+      this.setState({newWorksheetName: 'user-sampleworksheet'});
```
Comment (percyliang) [inline codalab/apps/web/static/js/worksheet/new_worksheet.jsx]: don't duplicate 'user-sampleworksheet' - this should appear once

File: codalab/apps/web/static/js/worksheet/new_worksheet.jsx [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+  render: function () {
+    var new_worksheet_name = (
+        <input type='text' id='new-worksheet-input' value={this.state.newWorksheetName} onChange={this.handleNameChange} ></input>
+      );
+    return (
+      <div>
+        <div id='new-worksheet'>
+          <span className='close' onClick={this.toggleNewWorksheet}>×</span>
+          <p className='pop-up-title'>New Worksheet</p>{new_worksheet_name}
```
Comment (percyliang) [inline codalab/apps/web/static/js/worksheet/new_worksheet.jsx]: Create New Worksheet

File: codalab/apps/web/static/js/worksheet/new_worksheet.jsx [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+          <p id='new-worksheet-message' className='pop-up-text'>Equivalent web terminal command:<br />
+            <span className='pop-up-command'>cl new {this.state.newWorksheetName}</span>
+          </p>
+          <div id='new-worksheet-button'>
+            <button className='pop-up-button' onClick={this.toggleNewWorksheet}>Cancel</button>
```
Comment (percyliang) [inline codalab/apps/web/static/js/worksheet/new_worksheet.jsx]: Swap the order of these two buttons.

**Comment by percyliang** [comment]
1) Make keyboard shortcuts work: 'Enter' should click 'Create', 'ESC' should click 'Cancel'.
2) Two buttons should look a bit more different...
3) Don't allow people to type non-worksheet characters (look at the regex).

File: codalab/apps/web/static/js/worksheet/new_worksheet.jsx (hunk 2 above)
Comment (ChengshuLi) [inline codalab/apps/web/static/js/worksheet/new_worksheet.jsx]: I also followed the Google Drive "create new folder" model. I think 'New Worksheet' conveys the meaning and is more concise as well. What do you think?

**Comment by ChengshuLi** [comment]
@percyliang  Take a look.

File: codalab/apps/web/static/js/worksheet/new_worksheet.jsx (hunk 2 above)
Comment (percyliang) [inline codalab/apps/web/static/js/worksheet/new_worksheet.jsx]: sounds good

File: codalab/apps/web/static/js/worksheet/worksheet_interface.jsx [hunk 4]
```diff
@@ -12,6 +12,7 @@ var Worksheet = React.createClass({
         return {
             ws: new WorksheetContent(this.props.uuid),
             version: 0,  // Increment when we refresh
+            escCount: 0, // Increment when the user presses esc keyboard short cut, a hack to allow esc shortcut to work
```
Comment (percyliang) [inline codalab/apps/web/static/js/worksheet/worksheet_interface.jsx]: 'short cut' is one word

File: codalab/apps/web/static/js/worksheet/new_worksheet.jsx [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        type='default'
+        handleClick={this.toggleNewWorksheet}
+      />
+    );
+    return (
+      <div>
+        <div id='new-worksheet'>
+          <span className='close' onClick={this.toggleNewWorksheet}>×</span>
+          <p className='pop-up-title'>New Worksheet</p>{new_worksheet_name}
+          <p id='new-worksheet-message' className='pop-up-text'>Equivalent web terminal command:<br />
```
Comment (percyliang) [inline codalab/apps/web/static/js/worksheet/new_worksheet.jsx]: This is kind of verbose, maybe replace with 'CodaLab> cl new ...' all on one line (same for the new run bundle button).

File: codalab/apps/web/static/js/worksheet/new_worksheet.jsx [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+            <span className='pop-up-command'>cl new {this.state.newWorksheetName}</span>
+          </p>
+          <div id='new-worksheet-button'>
+            {cancel_button}
+            {create_button}
+          </div>
+        </div>
+        <button onClick={this.toggleNewWorksheet}>Create New Worksheet</button>
```
Comment (percyliang) [inline codalab/apps/web/static/js/worksheet/new_worksheet.jsx]: Replace with 'New Worksheet' then?

**Comment by percyliang** [comment]
Final thing: can the focus be placed on the main body of the worksheet rather than the web terminal after the worksheet is created?

**Comment by ChengshuLi** [comment]
@percyliang  Have a look :)

File: codalab/apps/web/static/js/worksheet/new_worksheet.jsx [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        type='default'
+        handleClick={this.toggleNewWorksheet}
+      />
+    );
+    return (
+      <div>
+        <div id='new-worksheet'>
+          <span className='close' onClick={this.toggleNewWorksheet}>×</span>
+          <p className='pop-up-title'>New Worksheet</p>{new_worksheet_name}
+          <p id='new-worksheet-message' className='pop-up-text'>CodaLab>
```
Comment (percyliang) [inline codalab/apps/web/static/js/worksheet/new_worksheet.jsx]: add an extra space after CodaLab>

**Comment by percyliang** [comment]
We still need to get the username, right?

**Comment by ChengshuLi** [comment]
@percyliang Fixed.

File: codalab/apps/web/static/js/worksheet/new_worksheet.jsx [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+      return;
+    }
+    var command = 'cl new \"' + this.state.newWorksheetName + '\"';
+    response = $('#command_line').terminal().exec(command);
+    this.toggleNewWorksheet();
+  },
+
+  handleKeyDown: function(e) {
+    if (e.keyCode === 13) {
+      // enter shortcur
```
Comment (percyliang) [inline codalab/apps/web/static/js/worksheet/new_worksheet.jsx]: typo

**Comment by percyliang** [comment]
LGTM modulo typo

**Pull request merged**